karpenter-operator/controllers/karpenter/karpenter_controller.go

```diff
@@ -0,0 +1,328 @@
+package karpenter
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	hyperv1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	"github.com/openshift/hypershift/karpenter-operator/controllers/karpenter/assets"
+	supportassets "github.com/openshift/hypershift/support/assets"
+	"github.com/openshift/hypershift/support/upsert"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	karpenterFinalizer = "hypershift.openshift.io/karpenter-finalizer"
+	// userDataAMILabel is a label set in the userData secret generated for karpenter instances.
+	userDataAMILabel = "hypershift.openshift.io/ami"
+)
+
+var (
+	crdEC2NodeClass = supportassets.MustCRD(assets.ReadFile, "karpenter.k8s.aws_ec2nodeclasses.yaml")
+	crdNodePool     = supportassets.MustCRD(assets.ReadFile, "karpenter.sh_nodepools.yaml")
+	crdNodeClaim    = supportassets.MustCRD(assets.ReadFile, "karpenter.sh_nodeclaims.yaml")
+)
+
+type Reconciler struct {
+	ManagementClient          client.Client
+	GuestClient               client.Client
+	Namespace                 string
+	ControlPlaneOperatorImage string
+	upsert.CreateOrUpdateProvider
+}
+
+func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, managementCluster cluster.Cluster) error {
+	r.ManagementClient = managementCluster.GetClient()
+	r.GuestClient = mgr.GetClient()
+	r.CreateOrUpdateProvider = upsert.New(false)
+
+	// First install the CRDs so we can create a watch below.
+	mgr.GetClient()
```

Comment:
the return value is ignored, is this needed?

Reply:
no needed,  I dropped it. Thanks!